### ISSUE TITLE
Problems with TCP connection

### ISSUE THREAD

**Issue by spunky1a** [tomojitakasu/RTKLIB#153]
Hello,

following situation, I connect my M8T via Bluetooth with my phone, I send the raw data via TCP connection to RTKlib which fetches the correction data via NTRIP and smartphone fetches at the end by TCP, the corrected data as NMEA from. So far, everything works beautifully.

I close the connection in the app on my phone, but the connection to RTKlib is apparently not properly closed. A reconnection works, but I can not send raw data, since the port is probably blocked by the first connection.

This is not a problem of the app, the same behavior can I watch when I establish a connection with HyperTerminal, when disconnecting and reconnecting a new client at RTKlib appears.

So far only solution, I finish RTKlib with VNC and restart, then reconnect with the app.

Can you add an option, what close the 1st connection, if have the 2nd connection the same ip-adress like the 1st?

look here too: http://www.gpsforum.geospector.de/phpbb/viewtopic.php?f=4&t=3636

Thank you.

**Comment by linkaiqin** [comment]
Hi spunky1a 
I have the same problem with you.   If TCP connection closed by peer, the socket api recv() will return zero, but rtklib doesn't check the return value. To fix the issue, modify recv_nb() function in stream.c below.  
```
/* non-block receive ---------------------------------------------------------*/
static int recv_nb(socket_t sock, unsigned char *buff, int n)
{
    struct timeval tv={0};
    fd_set rs;
    FD_ZERO(&rs); FD_SET(sock,&rs);

    if (!select(sock+1,&rs,NULL,NULL,&tv)) return 0;
    return recv(sock,(char *)buff,n,0);
}
```
--->
```
/* non-block receive ---------------------------------------------------------*/
static int recv_nb(socket_t sock, unsigned char *buff, int n)
{
    struct timeval tv={0};
    fd_set rs;
    int len;
    
    FD_ZERO(&rs); FD_SET(sock,&rs);
    if (!select(sock+1,&rs,NULL,NULL,&tv)) return 0;
    if((len=recv(sock,(char *)buff,n,0))<=0) return -1; 
    return len;
}
```

**Comment by spunky1a** [comment]
oh great, thank you, it works perfect. :+1:

**Comment by tomojitakasu** [comment]
The problem is fixed in rtklib 2.4.3 b20.
